File: AST/AstValueFactory.cpp
// For generating constants.
#define AST_STRING_CONSTANTS(F)                 \
  F(anonymous, "anonymous")                     \
  F(anonymous_function, "(anonymous function)") \
  F(arguments, "arguments")                     \
  F(as, "as")                                   \
  F(async, "async")                             \
  F(await, "await")                             \
  F(bigint, "bigint")                           \
  F(boolean, "boolean")                         \
  F(computed, "<computed>")                     \
  F(dot_brand, ".brand")                        \
  F(constructor, "constructor")                 \
  F(default, "default")                         \
  F(done, "done")                               \
  F(dot, ".")                                   \
  F(dot_default, ".default")                    \
  F(dot_for, ".for")                            \
  F(dot_generator_object, ".generator_object")  \
  F(dot_iterator, ".iterator")                  \
  F(dot_promise, ".promise")                    \
  F(dot_result, ".result")                      \
  F(dot_switch_tag, ".switch_tag")              \
  F(dot_catch, ".catch")                        \
  F(empty, "")                                  \
  F(eval, "eval")                               \
  F(from, "from")                               \
  F(function, "function")                       \
  F(get, "get")                                 \
  F(get_space, "get ")                          \
  F(length, "length")                           \
  F(let, "let")                                 \
  F(meta, "meta")                               \
  F(name, "name")                               \
  F(native, "native")                           \
  F(new_target, ".new.target")                  \
  F(next, "next")                               \
  F(number, "number")                           \
  F(object, "object")                           \
  F(of, "of")                                   \
  F(private_constructor, "#constructor")        \
  F(proto, "__proto__")                         \
  F(prototype, "prototype")                     \
  F(return, "return")                           \
  F(set, "set")                                 \
  F(set_space, "set ")                          \
  F(string, "string")                           \
  F(symbol, "symbol")                           \
  F(target, "target")                           \
  F(this, "this")                               \
  F(this_function, ".this_function")            \
  F(throw, "throw")                             \
  F(undefined, "undefined")                     \
  F(value, "value")

#define F(name, str)                           \
  const AstRawString* name##_string() const {  \
    return string_constants_->name##_string(); \
  }
  AST_STRING_CONSTANTS(F)
#undef F

class AstValueFactory {
  public:
    AstValueFactory(Zone* zone, const AstStringConstants* string_constants,
                    uint64_t hash_seed)
        : string_table_(string_constants->string_table()),
          strings_(nullptr),
          strings_end_(&strings_),
          cons_strings_(nullptr),
          cons_strings_end_(&cons_strings_),
          string_constants_(string_constants),
          empty_cons_string_(nullptr),
          zone_(zone),
          hash_seed_(hash_seed) {
      std::fill(one_character_strings_,
                one_character_strings_ + arraysize(one_character_strings_),
                nullptr);
      empty_cons_string_ = NewConsString();
    }
    const AstRawString* GetOneByteString(Vector<const uint8_t> literal) {
      return GetOneByteStringInternal(literal);
    }
    AstRawString* GetOneByteStringInternal(
      Vector<const uint8_t> literal) {
      if (literal.length() == 1 && literal[0] < kMaxOneCharStringValue) {
        int key = literal[0];
        if (V8_UNLIKELY(one_character_strings_[key] == nullptr)) {
          uint32_t hash_field = StringHasher::HashSequentialString<uint8_t>(
              literal.begin(), literal.length(), hash_seed_);
          one_character_strings_[key] = GetString(hash_field, true, literal);
        }
        return one_character_strings_[key];
      }
      uint32_t hash_field = StringHasher::HashSequentialString<uint8_t>(
          literal.begin(), literal.length(), hash_seed_);
      return GetString(hash_field, true, literal);
    }
    AstRawString* GetString(uint32_t hash_field, bool is_one_byte,
                                         Vector<const byte> literal_bytes) {
      // literal_bytes here points to whatever the user passed, and this is OK
      // because we use vector_compare (which checks the contents) to compare
      // against the AstRawStrings which are in the string_table_. We should not
      // return this AstRawString.
      AstRawString key(is_one_byte, literal_bytes, hash_field);
      base::HashMap::Entry* entry = string_table_.LookupOrInsert(&key, key.Hash());
      if (entry->value == nullptr) {
        // Copy literal contents for later comparison.
        int length = literal_bytes.length();
        byte* new_literal_bytes = zone_->NewArray<byte>(length);
        memcpy(new_literal_bytes, literal_bytes.begin(), length);
        AstRawString* new_string = new (zone_) AstRawString(
            is_one_byte, Vector<const byte>(new_literal_bytes, length), hash_field);
        CHECK_NOT_NULL(new_string);
        AddString(new_string);
        entry->key = new_string;
        entry->value = reinterpret_cast<void*>(1);
      }
      return reinterpret_cast<AstRawString*>(entry->key);
    }
  private:
    // All strings are copied here, one after another (no zeroes inbetween).
    base::CustomMatcherHashMap string_table_;

    // We need to keep track of strings_ in order since cons strings require their
    // members to be internalized first.
    AstRawString* strings_;
    AstRawString** strings_end_;
    AstConsString* cons_strings_;
    AstConsString** cons_strings_end_;

    // Holds constant string values which are shared across the isolate.
    const AstStringConstants* string_constants_;
    const AstConsString* empty_cons_string_;

    // Caches one character lowercase strings (for minified code).
    static const int kMaxOneCharStringValue = 128;
    AstRawString* one_character_strings_[kMaxOneCharStringValue];

    Zone* zone_;

    uint64_t hash_seed_;
}

template <typename schar>
uint32_t StringHasher::HashSequentialString(const schar* chars, int length,
                                            uint64_t seed) {
  // Check whether the string is a valid array index. In that case, compute the
  // array index hash. It'll fall through to compute a regular string hash from
  // the start if it turns out that the string isn't a valid array index.
  if (IsInRange(length, 1, String::kMaxArrayIndexSize)) {
    if (IsDecimalDigit(chars[0]) && (length == 1 || chars[0] != '0')) {
      uint32_t index = chars[0] - '0';
      int i = 1;
      do {
        if (i == length) {
          return MakeArrayIndexHash(index, length);
        }
      } while (TryAddIndexChar(&index, chars[i++]));
    }
  } else if (length > String::kMaxHashCalcLength) {
    return GetTrivialHash(length);
  }

  // Non-array-index hash.
  DCHECK_LE(0, length);
  DCHECK_IMPLIES(0 < length, chars != nullptr);
  uint32_t running_hash = static_cast<uint32_t>(seed);
  const schar* end = &chars[length];
  while (chars != end) {
    running_hash = AddCharacterCore(running_hash, *chars++);
  }

  return (GetHashCore(running_hash) << String::kHashShift) |
         String::kIsNotArrayIndexMask;
}

uint32_t StringHasher::MakeArrayIndexHash(uint32_t value, int length) {
  // For array indexes mix the length into the hash as an array index could
  // be zero.
  value <<= String::ArrayIndexValueBits::kShift;
  value |= length << String::ArrayIndexLengthBits::kShift;
  return value;
}

template <typename Char>
bool TryAddIndexChar(uint32_t* index, Char c) {
  if (!IsDecimalDigit(c)) return false;
  int d = c - '0';
  if (*index > 429496729U - ((d + 3) >> 3)) return false;
  *index = (*index) * 10 + d;
  return true;
}

File: ast/AstValueFactory.js
import StringHasher from './StringHasher';
import AstRawString from "./AstRawString";
import { ZoneObject } from './Ast';

const kMaxOneCharStringValue = 128;

export default class AstValueFactory {
  constructor() {
    this.one_character_strings_ = new Array(kMaxOneCharStringValue).fill(null);
    this.string_table_ = [];
    // this.hash_seed_ = BigInt(15853730874361889590); 
    this.hash_seed_ = 1704808181;
    this.string_end_ = [];
    this.cons_string_ar = [];
    this.empty_cons_string_ = this.NewConsString();
  }

  dot_string() { return '.'; }
  eval_string() { return 'eval'; }
  name_string() { return 'name'; }
  constructor_string() { return 'constructor'; }
  arguments_string() { return 'arguments'; }
  dot_result_string() { return '.result'; }
  dot_generator_object_string() { return '.generator_object'; }
  dot_switch_tag_string() { return '.switch_tag'; }
  prototype_string() { return 'prototype'; }
  proto_string() { return '__proto__'; }
  empty_string() { return ''; }
  new_target_string() { return '.new.target'; }
  this_string() { return 'this'; }
  this_function_string() { return '.this_function'; }
  default_string() { return 'default'; }
  dot_default_string() { return '.default'; }
  get_space_string() { return 'get'; }
  set_space_string() { return 'set'; }
  native_string() { return 'native'; }
  of_string() { return 'of'; }
  computed_string() { return '<computed>'; }
  meta_string() { return 'meta'; }
  dot_catch_string() { return '.catch'; }

  empty_cons_string() { return this.empty_cons_string_; }
  NewConsString(str1 = null, str2 = null) {
    let new_string = new AstConsString();
    this.AddConsString(new_string);
    return new_string.AddString(str1).AddString(str2);
  }
  AddConsString(string) {
    /**
     * 这里用二维指针实现的一个链表
     * JS直接用数组代替了
     */
    this.cons_string_ar.push(string);
  }
  /**
   * 源码中参数是一个vector容器 这里直接用字符串代替了
   * @param {Vector<const uint8_t>} literal 
   */
  GetOneByteString(literal) {
    return this.GetOneByteStringInternal(literal);
  }
  GetOneByteStringInternal(literal) {
    // 单字符特殊处理 有一个映射表单独缓存了所有key
    if (literal.length === 1 && literal[0].charCodeAt() < kMaxOneCharStringValue) {
      let key = literal[0].charCodeAt();
      /**
       * V8_UNLIKELY
       * 单字符变量第一次出现会进这里
       * one_character_strings_是一个长度为128的Vector 保存每一个字符生成的hash值
       */
      if (this.one_character_strings_[key] === null) {
        let hash_field = StringHasher.HashSequentialString(literal, 1, this.hash_seed_);
        this.one_character_strings_[key] = this.GetString(hash_field, true, literal);
      }
      return this.one_character_strings_[key];
    }
    let hash_field = StringHasher.HashSequentialString(literal, literal.length, this.hash_seed_);
    // 最后返回的必须是无符号整数
    hash_field >>>= 0;
    return this.GetString(hash_field, true, literal);
  }
  GetString(hash_field, is_one_byte, literal_bytes) {
    let key = new AstRawString(is_one_byte, literal_bytes, hash_field);
    // 这里实际传的key是对象的指针
    let entry = this.LookupOrInsert(key, key.Hash());
    /**
     * 源码由HashMap返回
     * key是一个void*、value是对应的hash值
     */
    if (entry.value === null) {
      // 在指定的内存上初始化对象 对JS来说毫无意义
      // int length = literal_bytes.length();
      // byte* new_literal_bytes = zone_->NewArray<byte>(length);
      // memcpy(new_literal_bytes, literal_bytes.begin(), length);
      // AstRawString* new_string = new (zone_) AstRawString(is_one_byte, Vector<const byte>(new_literal_bytes, length), hash_field);
      this.AddString(key);
      // entry->key = new_string;
      // entry->value = reinterpret_cast<void*>(1);
      /**
       * C++中两个地址相同的对象全等
       * 所以上面的代码value直接用一个给定指针代替
       * JS不存在这种科技 因此用Hash值作为标示判定对象的相等
       */
      entry.key = key;
      entry.value = key.Hash();
    }
    // 源码将void*强转返回 => reinterpret_cast<AstRawString*>
    return entry.key;
  }
  /**
   * 
   * @param {AstRawString*} key 源码是指针 这里只能用对象代替
   * @param {Number} value Hash值
   */
  LookupOrInsert(key, hash) {
    let tar = this.string_table_.find(v => v.hash === hash);
    if (!tar) {
      let result = {
        key: null,
        value: null
      };
      this.string_table_.push(result);
      return result;
    }
    return tar;
  }
  AddString(string) {
    /**
     * 二维指针 搞屁
     */
    this.string_end_.push(string);
    return string;
  }

  Internalize() {}
}

class AstConsString extends ZoneObject {
  constructor() {
    super();
    this.next_ = null;
    // 链表
    this.segment_ = {
      string: null, 
      next: null,
    };
  }
  IsEmpty() {
    return this.segment_.string === null;
  }
  // 没有指针 实现起来有点变扭。。。
  AddString(s) {
    if(s === null) return this;
    if(!this.IsEmpty()) {
      let tmp = {
        string: s, 
        next: this.segment_,
      };
      this.segment_ = tmp;
    }
    this.segment_.string = s;
    return this;
  }
}